Two small QoL improvements

r618 opened the pull request:
- corrected [ExecuteInEditMode] / [ExecuteAlways] support with lazy init of scheduler where Awake is not called in Editor
- allow BeginWindow to return/modify also ImRect from state - useful for saving/restoring window's position/size

Commit:
fix [ExecuteAlways]: scheduler lazy init moved to `OnEnable` -

`Awake` is *not* called consistently enough for others [ExecuteAlways]/[ExecuteInEditMode] Updates.., Having scheduler in OnEnable seems to be enough for now, but needs whole lazy init currently in Awake to be updated/moved...

Commit:
[Window] make rect `ref` to make window size/position known

useful for saving/restoring window state after e.g. modified by dragging/resizing

tonycoculuzzi commented on Runtime/Scripts/IO/UGUI/ImuiUnityGUIBackend.cs:
```diff
@@ -122,6 +121,8 @@ protected override void OnEnable()
         {
             base.OnEnable();
 
+            scheduler ??= GraphicsSettings.currentRenderPipeline ? new ImuiScriptableRenderingScheduler() : new ImuiBuiltinRenderingScheduler();
+
```
This is a good change, for compatibility's sake.

Commit:
+ opt-in Input System mouse support for Game View Simulator

adds opt-in editor path for the backend that allows desktop mouse input to drive touches when using Game View Simulator

Commit:
- remove `simulateTouchWithMouseInEditor` this should work regardless of GameView mode .(..)

Commit:
FIX: MenuBar not showing menu on touches (e.g. Simulator mode in GameView)

it used Mouse.current.position which is not used in Simulator mode, for raycasting/OnPointerDown

Commit:
FIX: Restore Imui input in GameView Simulator

remove now not needed parts of the previous attempt

Commit:
Fix Imui pointer input under the Device Simulator

ImUnityInputWrapper.MousePosition returned (0,0) while hovering in the
   Simulator. The Simulator disables Mouse.current and reports a hovering
   touch with phase==Ended, which the mouse and active-touch paths both
   reject. Add Pointer.current (the common Mouse/Pen/Touchscreen base) as a
   final position fallback so hover tracks the cursor. Purely additive: the
   existing mouse/active-touch paths are unchanged when either is available.

Commit:
Fix Imui pointer input under the Device Simulator #2

ImuiUnityGUIBackend re-sampled the pointer position at queue-drain time
   instead of when the event occurred. Because the queue drains one event
   per frame, a touch that lifted (tap) or moved (drag) by drain time had
   its Down/Click attributed to the wrong position/control. Store the
   position alongside each queued event (QueuedMouseEvent) so events are
   self-describing. Also drop the IsTouchBeganThisFrame() guard on the
   deferred Move event, since that began-edge detection is itself
   unreliable in the Simulator; a Move is now enqueued on every touch-down.

Commit:
FIX lazy shader/material loading w/ [ExecuteAlways]

(became immediately visible with project in Linear mode)

vape commented on Editor/Scripts/ImuiUnityGUIBackend.cs:
This should be just discarded

vape commented on Runtime/Scripts/Controls/ImMenuBar.cs:
```diff
@@ -52,6 +52,11 @@ public static void BeginMenuBar(ImGui gui, uint id, ImRect rect, int order)
             gui.Layout.Push(ImAxis.Horizontal, rect);
             gui.Canvas.PushOrder(order);
 
+            if (!gui.WindowManager.IsDrawingWindow())
+            {
+                gui.RegisterRaycastTarget(rect);
+            }
+
```
This really should be a `gui.RegisterControl(id, rect)` call

vape commented on Runtime/Scripts/Controls/ImWindow.cs:
There is an API for querying current window state, which includes current rect:

```
if (gui.WindowManager.TryGetDrawingWindowId(out var windowId))
{
    ref var state = ref gui.WindowManager.GetWindowState(windowId);
    var rectToSave = state.Rect;
    // ...
}
```

I don't really like the idea of further complicating BeginWindow signature with ref parameter

vape commented on Runtime/Scripts/IO/UGUI/ImuiUnityGUIBackend.cs:
```diff
@@ -47,7 +59,19 @@ public enum ScalingMode
+        {
+            get
+            {
+                // Created lazily rather than in Awake: after an assembly reload [ExecuteAlways]
+                // components only get OnEnable, and with Enter Play Mode Options they get neither
+                // Awake nor OnEnable when entering play — an Awake-only setup leaves the canvas
+                // rebuild binding base.defaultMaterial, which skips the color-space conversion
+                // this backend's shader performs in Linear projects.
```
Although I get the justification and I'm okay with the changes, this comment is wrong and should be adjusted. OnEnable is called always, even with domain and scene reload disabled.

r618 replied:
this comment was not present in original PR commit, but it is now at [develop's ](r618:develop) latest
. i made more changes in the fork/branch related to Linear vs. sRGB and mainly Simulator GameView mode, where more stuff didn't work OOTB
will you be ok closing this PR and me submitting clean branch just with these two original changes - or advise otherwise
( I hope to open PR with Simulator related stuff hopefully later, when it's cleaned up more (also some of your other comments are related to this..)), sorry for confusion

btw your other comments will be still helpful regardless so no harm in closing this; the real content of this PR will be really just the lazy initialization, will open new PR later..;

vape commented on Runtime/Scripts/IO/Utility/ImUnityInputWrapper.cs:
```diff
@@ -24,14 +77,73 @@ public static bool IsTouchBeganThisFrame()
             var touches = touchscreen.touches;
             for (int i = 0; i < touches.Count; ++i)
             {
-                if (touches[i].phase.value == UnityEngine.InputSystem.TouchPhase.Began)
+                if (WasTouchPressedThisFrame(touches[i]))
                 {
                     return true;
                 }
             }
 
+            return WasTouchPressedThisFrame(touchscreen.primaryTouch);
+        }
```
Its not referenced anywhere anymore and should be removed

vape commented on Runtime/Scripts/IO/UGUI/ImuiUnityGUIBackend.cs:
```diff
@@ -148,6 +164,33 @@ protected override void OnEnable()
+            };
+            return DefaultMaterial;
+        }
+
+        protected override void OnDisable()
+        {
+            base.OnDisable();
+        }
+
```
Probably left after debugging something, should be removed

vape reviewed (changes requested):
Hey, I finally got time to look into it. Overall, I think changes are useful and would like to merge them, but before that a few comments should be addressed.

r618 commented:
opened https://github.com/vape/Imui/pull/24 for 1st change
will close this
Thanks !
